query-node/mappings/src/content/utils.ts

```diff
@@ -0,0 +1,484 @@
+  CuratorGroup,
+  Channel,
+  ChannelCategory,
+  Video,
+  VideoCategory,
+
+  // secondary entities
+  Language,
+  License,
+  VideoMediaEncoding,
+  VideoMediaMetadata,
+
+  // asset
+  DataObjectOwner,
+  DataObjectOwnerMember,
+  DataObjectOwnerChannel,
+  DataObject,
+  LiaisonJudgement,
+  AssetAvailability,
+} from 'query-node'
+
+// Joystream types
+import {
+  ChannelId,
+  ContentParameters,
+  NewAsset,
+  ContentActor,
+} from '@joystream/types/augment'
+
+/*
+  Asset either stored in storage or describing list of URLs.
+*/
+type AssetStorageOrUrls = DataObject | string[]
+
+/*
+  Type guard differentiating asset stored in storage from asset describing a list of URLs.
+*/
+function isAssetInStorage(dataObject: AssetStorageOrUrls): dataObject is DataObject {
+  if (Array.isArray(dataObject)) {
+    return false
+  }
+
+  return true
+}
+
+export interface IReadProtobufArguments {
+  metadata: Bytes
+  db: DatabaseManager
+  blockNumber: number
+}
+
+export interface IReadProtobufArgumentsWithAssets extends IReadProtobufArguments {
+  assets: NewAsset[] // assets provided in event
+  contentOwner: typeof DataObjectOwner
+}
+
+/*
+  Reads information from the event and protobuf metadata and constructs changeset that's fit to be used when saving to db.
+*/
+export async function readProtobuf<T extends ChannelCategory | VideoCategory>(
+  type: T,
+  parameters: IReadProtobufArguments,
+): Promise<Partial<T>> {
+  // true option here is crucial, it indicates that we want just the underlying bytes (by default it will also include bytes encoding the length)
+  const metaU8a = parameters.metadata.toU8a(true);
+
+  // process channel category
+  if (type instanceof ChannelCategory) {
+    return ChannelCategoryMetadata.deserializeBinary(metaU8a).toObject() as Partial<T>
+  }
+
+  // process video category
+  if (type instanceof VideoCategory) {
+    return VideoCategoryMetadata.deserializeBinary(metaU8a).toObject() as Partial<T>
+  }
+
+  // this should never happen
+  logger.error('Not implemented metadata type', {type})
+  throw `Not implemented metadata type`
+}
+
+/*
+  Reads information from the event and protobuf metadata and constructs changeset that's fit to be used when saving to db.
+  In addition it handles any assets associated with the metadata.
+*/
+
+/*
+export async function readProtobufWithAssets(
+  type: Channel | Video,
+  parameters: IReadProtobufArgumentsWithAssets,
+): Promise<Partial<typeof type>> {
+*/
+export async function readProtobufWithAssets<T extends Channel | Video>(
+  type: T,
+  parameters: IReadProtobufArgumentsWithAssets,
+): Promise<Partial<T>> {
+  // true option here is crucial, it indicates that we want just the underlying bytes (by default it will also include bytes encoding the length)
+  const metaU8a = parameters.metadata.toU8a(true);
+
+  // process channel
+  if (type instanceof Channel) {
+    const meta = ChannelMetadata.deserializeBinary(metaU8a)
+    const metaAsObject = meta.toObject()
```

Comment:
The approach you are taking here by converting to a JSON object is loosing important information about whether the field was set or not. It is not sufficient to later check if metaAsObject.someField === undefined.

Reply:
This approach has been reworked to check if value was trully set (using `has...` prefixed functions like `hasCoverPhoto()`).